mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -2319,3 +2371,167 @@ mlir::rock::removeUpperDims(OpBuilder &b, ArrayAttr transformAttrs,
+          for (auto [lowDim, subDimStride, param] : llvm::zip(
+                   trAttr.getLowerDims(), subDimStrides, trAttr.getParams())) {
+            int64_t upperDim = trAttr.getUpperDims()[0];
+            if (currSubDimInfo.contains(upperDim)) {
+              for (const SubDimInfo &sdInfo : currSubDimInfo.at(upperDim)) {
+                if (sdInfo.stride >= subDimStride * param) {
+                  LLVM_DEBUG(llvm::dbgs()
+                             << "No overlap: stride of analyzed dim is larger "
+                                "than new subdim stride.\n");
+                } else if (sdInfo.stride * sdInfo.size <= subDimStride) {
+                  LLVM_DEBUG(llvm::dbgs()
+                             << "No overlap: stride of new subdim stride is "
+                                "larger than the analyzed dim.\n");
+                } else {
+                  // New sizes and strides for newly annotated subdims
+                  int64_t newSize;
+                  int64_t newStride;
+                  int64_t maxStrideSubDim = subDimStride * param;
+                  // Overlap on the right side of annotated subdim
+                  if (sdInfo.stride * sdInfo.size >= maxStrideSubDim) {
+                    int64_t rhsBoundForAnnotate =
+                        std::max(sdInfo.stride, subDimStride);
+                    newSize = maxStrideSubDim / rhsBoundForAnnotate;
+                    newStride = rhsBoundForAnnotate / subDimStride;
+                  }
+                  // The whole of annotatedSubDim is within the newly created
+                  // lowDim
+                  else if (sdInfo.stride >= subDimStride) {
+                    newSize = sdInfo.size;
+                    newStride = sdInfo.stride / subDimStride;
+                  }
+                  // Overlap on the left side of annotated subdim
+                  else {
+                    int64_t maxStrideRemovedSubDim =
+                        sdInfo.stride * sdInfo.size;
+                    newSize = maxStrideRemovedSubDim / subDimStride;
+                    newStride = 1;
+                  }
+                  LLVM_DEBUG(llvm::dbgs() << "creating subDim of <size="
+                                          << newSize << ",stride=" << newStride
+                                          << "> @ " << lowDim << "\n");
+                  if (newSize > 1) {
+                    nextSubDimInfo[lowDim].push_back({newSize, newStride});
+                  }
+                }
+              }
+            }
+          }
+        } break;
+        case TransformType::Unmerge: {
+          SmallVector<int64_t> subDimStrides = getStrides(trAttr.getParams());
+          int64_t lowDim = trAttr.getLowerDims()[0];
+          for (size_t subDim = 0; subDim < trAttr.getParams().size();
+               subDim++) {
+            int64_t upperDim = trAttr.getUpperDims()[subDim];
+            if (currSubDimInfo.contains(upperDim)) {
```

Comment:
I thought it currSubDimInfo.contains(upperDim) should always be true, when can this be false?